### PR TITLE
add --effort 1 and 2

### PR DESCRIPTION
**Pull request by jonsneyers**
Adds faster encode options, mostly useful for lossless mode (since it is reversible and you can always do something better later; in lossy mode that's not really the case).

New speeds are called `thunder` and `lightning`, because we ran out of animals (the peregrine falcon is the fastest animal that exists), and correspond to `--effort 2` and `--effort 1`, respectively.

For VarDCT mode, what used to be `falcon` is now called `thunder`, and the new `falcon` is a bit slower but it gets a better density, making it have a more attractive trade-off (the gap between `cheetah` and `falcon` was quite large).

For Modular mode, existing speed settings have not changed; `thunder` and `lightning` are just new settings. At the moment they are almost the same (Gradient predictor with a fixed tree), the only difference being in the context clustering (which seems to help more for decode speed than for encode speed; perhaps later something actually faster can be done there).

Some example timings: (not very accurate; I used `--num_threads=0 --encode_reps=10 --decode_reps=30` in a 4-core VirtualBox Debian on a macbook, so only SSE4 got used, and I didn't really pay attention to cpu throttling effects)

Before:
```
ClassA_8bit_BIKE_2048x2560_8b_RGB.ppm.png
Compr                Input    Compr            Compr       Compr  Decomp  Butteraugli                                             
Method              Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
----------------------------------------------------------------------------------------------------------------------------------
jxl:cheetah        5242880  1048906    1.60050354004   1  26.837  23.857   2.65753889   0.81365609515  1.3022594606562341        0
jxl:falcon         5242880  1180185    1.80081939697   1  37.762  24.221   1.75961113   0.79067702827  1.4238665292565629        0
jxl:m:falcon       5242880  7595698   11.59011535645   1  12.340   7.048   0.00000000   0.00000000000  0.0000000000000000        0
```

After:
```
ClassA_8bit_BIKE_2048x2560_8b_RGB.ppm.png
Compr                   Input    Compr            Compr       Compr  Decomp  Butteraugli                                             
Method                 Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
-------------------------------------------------------------------------------------------------------------------------------------
jxl:cheetah           5242880  1048906    1.60050354004   1  28.414  24.160   2.65753889   0.81365609515  1.3022594606562341        0
jxl:falcon            5242880  1088227    1.66050262451   1  36.011  23.570   1.75961113   0.79067702827  1.3129212805901462        0
jxl:thunder           5242880  1180185    1.80081939697   1  37.735  25.222   1.75961113   0.79067702827  1.4238665292565629        0
jxl:m:falcon          5242880  7595698   11.59011535645   1  12.327   7.053   0.00000000   0.00000000000  0.0000000000000000        0
jxl:m:thunder         5242880  8099161   12.35833892822   1  18.767  15.773   0.00000000   0.00000000000  0.0000000000000000        0
jxl:m:lightning       5242880  8254977   12.59609527588   1  18.412  17.208   0.00000000   0.00000000000  0.0000000000000000        0
```

Before:
```
ClassD_APPLE_BasketBallScreen_2560x1440p_60_8b_sRGB_444_000.ppm.png
Compr                Input    Compr            Compr       Compr  Decomp  Butteraugli                                             
Method              Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
----------------------------------------------------------------------------------------------------------------------------------
jxl:cheetah        3686400   677662    1.47062065972   1  27.767  28.435   2.64845109   0.89112170499  1.3105019896857066        0
jxl:falcon         3686400   773637    1.67889973958   1  53.659  28.511   2.23516011   0.72212176554  1.2123700441151199        0
jxl:m:falcon       3686400  1884005    4.08855251736   1  13.434   7.430   0.00000000   0.00000000000  0.0000000000000000        0
```

After:
```
ClassD_APPLE_BasketBallScreen_2560x1440p_60_8b_sRGB_444_000.ppm.png
Compr                   Input    Compr            Compr       Compr  Decomp  Butteraugli                                             
Method                 Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
-------------------------------------------------------------------------------------------------------------------------------------
jxl:cheetah           3686400   677662    1.47062065972   1  28.470  28.365   2.64845109   0.89112170499  1.3105019896857066        0
jxl:falcon            3686400   698540    1.51592881944   1  45.908  26.572   2.23516011   0.72212176554  1.0946851955324990        0
jxl:thunder           3686400   773637    1.67889973958   1  50.820  27.134   2.23516011   0.72212176554  1.2123700441151199        0
jxl:m:falcon          3686400  1884005    4.08855251736   1  12.759   7.431   0.00000000   0.00000000000  0.0000000000000000        0
jxl:m:thunder         3686400  1832179    3.97608289931   1  20.705  16.674   0.00000000   0.00000000000  0.0000000000000000        0
jxl:m:lightning       3686400  2632232    5.71230902778   1  18.789  17.205   0.00000000   0.00000000000  0.0000000000000000        0
```

Before:
```
ClassA_8bit_WOMAN_2048x2560_8b_RGB.ppm.png
Compr                Input    Compr            Compr       Compr  Decomp  Butteraugli                                             
Method              Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
----------------------------------------------------------------------------------------------------------------------------------
jxl:cheetah        5242880   969626    1.47953186035   1  30.637  24.364   2.16948843   0.79623684190  1.1780577759831152        0
jxl:falcon         5242880  1133865    1.73014068604   1  37.983  24.685   1.80241621   0.84204727124  1.4568602435453724        0
jxl:m:falcon       5242880  7420429   11.32267608643   1  12.614   7.260   0.00000000   0.00000000000  0.0000000000000000        0
```

After:
```
ClassA_8bit_WOMAN_2048x2560_8b_RGB.ppm.png
Compr                   Input    Compr            Compr       Compr  Decomp  Butteraugli                                             
Method                 Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
-------------------------------------------------------------------------------------------------------------------------------------
jxl:cheetah           5242880   969626    1.47953186035   1  30.237  24.489   2.16948843   0.79623684190  1.1780577759831152        0
jxl:falcon            5242880  1032801    1.57592926025   1  36.114  22.894   1.80241621   0.84204727124  1.3270069332715131        0
jxl:thunder           5242880  1133865    1.73014068604   1  37.397  24.354   1.80241621   0.84204727124  1.4568602435453724        0
jxl:m:falcon          5242880  7420429   11.32267608643   1  12.290   7.012   0.00000000   0.00000000000  0.0000000000000000        0
jxl:m:thunder         5242880  7961832   12.14879150391   1  18.993  15.753   0.00000000   0.00000000000  0.0000000000000000        0
jxl:m:lightning       5242880  8088296   12.34176025391   1  17.581  16.898   0.00000000   0.00000000000  0.0000000000000000        0
```

Before:
```
ClassA_8bit_CAFE_2048x2560_8b_RGB.ppm.png
Compr                Input    Compr            Compr       Compr  Decomp  Butteraugli                                             
Method              Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
----------------------------------------------------------------------------------------------------------------------------------
jxl:cheetah        2048000   862723    3.37001171875   1  19.263  17.829   2.29643035   0.74632513037  2.5151244353587807        0
jxl:falcon         2048000   974050    3.80488281250   1  18.864  18.261   1.59732199   0.69378372361  2.6397657655436695        0
jxl:m:falcon       2048000  3750966   14.65221093750   1   3.683   6.304   0.00000000   0.00000000000  0.0000000000000000        0
```

After:
```
ClassA_8bit_CAFE_2048x2560_8b_RGB.ppm.png
Compr                   Input    Compr            Compr       Compr  Decomp  Butteraugli                                             
Method                 Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
-------------------------------------------------------------------------------------------------------------------------------------
jxl:cheetah           2048000   862723    3.37001171875   1  19.838  21.271   2.29643035   0.74632513037  2.5151244353587807        0
jxl:falcon            2048000   922924    3.60517187500   1  22.161  18.156   1.59732199   0.69378372361  2.5012095676799193        0
jxl:thunder           2048000   974050    3.80488281250   1  24.017  20.861   1.59732199   0.69378372361  2.6397657655436695        0
jxl:m:falcon          2048000  3750966   14.65221093750   1   4.210   6.336   0.00000000   0.00000000000  0.0000000000000000        0
jxl:m:thunder         2048000  3934312   15.36840625000   1   4.963  12.738   0.00000000   0.00000000000  0.0000000000000000        0
jxl:m:lightning       2048000  3979422   15.54461718750   1   4.681  13.534   0.00000000   0.00000000000  0.0000000000000000        0
```

### CODE REVIEW

File: lib/jxl/enc_frame.cc [hunk 1]
```diff
@@ -966,7 +966,7 @@ class LossyFrameEncoder {
     for (size_t i = 0; i < enc_state_->progressive_splitter.GetNumPasses();
          i++) {
       // No coefficient reordering in Falcon mode.
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:968]: Comment is outdated :)

**Review by veluca93 — CHANGES_REQUESTED** [review]
We probably should also add tests that skip the fast path for a gradient-only tree :)

File: lib/jxl/enc_frame.cc (hunk 1 above)
Comment (jonsneyers) [reply lib/jxl/enc_frame.cc]: Better?

**Comment by jonsneyers** [comment]
> We probably should also add tests that skip the fast path for a gradient-only tree :)

Added a test.

**Comment by jonsneyers** [comment]
@googlebot I signed it!

File: lib/jxl/enc_frame.cc [hunk 2]
```diff
@@ -965,8 +965,8 @@ class LossyFrameEncoder {
         enc_state_->progressive_splitter.GetNumPasses());
     for (size_t i = 0; i < enc_state_->progressive_splitter.GetNumPasses();
          i++) {
-      // No coefficient reordering in Falcon mode.
-      if (enc_state_->cparams.speed_tier != SpeedTier::kFalcon) {
+      // No coefficient reordering in Falcon or faster.
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:968]: Still wrong :)

**Review by veluca93 — CHANGES_REQUESTED** [review]

File: lib/jxl/enc_frame.cc (hunk 2 above)
Comment (jonsneyers) [reply lib/jxl/enc_frame.cc:968]: Lol can you suggest a fix? We're not doing coeff reordering in falcon, are we?
(if so, then is it so fast that we don't want to disable it?)

Comment (veluca93) [reply lib/jxl/enc_frame.cc:968]: You switched the condition to >, so no coefficient reordering in *faster than* falcon, no?

Comment (jonsneyers) [reply lib/jxl/enc_frame.cc:968]: No, that test is testing "if slower than falcon, do something". The speed_tier in the code is actual speed (faster is higher number), so it's the inverse of the --effort in cjxl.

The difference between falcon and thunder (for vardct) is that thunder (like the old falcon) does `ClusteringType::kFastest` while the new falcon uses `ClusteringType::kFast`.

Comment (veluca93) [reply lib/jxl/enc_frame.cc:968]: Ah, ok :D I guess we could also do coefficient reordering for falcon now, but meh

**Review by veluca93 — APPROVED** [review]

**Pull request merged**